client.go

```diff
@@ -115,16 +140,38 @@ func (c *Client) handleResponse(apiResponse *http.Response, response interface{}
 		var errorResponse ErrorResponse
 		err := json.Unmarshal(body, &errorResponse)
 		if err != nil {
-			return fmt.Errorf("api error(%v): no body", apiResponse.StatusCode)
+			return NewMapboxError(apiResponse.StatusCode, "")
 		}
 
-		return fmt.Errorf("api error(%v): %v", apiResponse.StatusCode, errorResponse.Message)
+		// If rate limited, hold off till the next X-Rate-Limit-Reset
+		if apiResponse.StatusCode == 429 {
+			resetUnix, err := strconv.Atoi(apiResponse.Header.Get("X-Rate-Limit-Reset"))
+			if err == nil {
+				c.rateLimits[rateLimit] = time.Unix(int64(resetUnix), 0)
+			}
+		}
+		return NewMapboxError(apiResponse.StatusCode, errorResponse.Message)
 	}
 
-	//convert to response
+	// convert to response
 	if err := json.Unmarshal(body, &response); err != nil {
 		return fmt.Errorf("failed to read body. %w", err)
 	}
 
 	return nil
 }
+
+func (c *Client) checkRateLimit(rl RateLimit) error {
+	reset := c.rateLimits[rl]
```

Comment:
I think you have a race condition here (read on line 165, write on line 150). Your test probably catches this too if you run with `-race` flag I would think.

You'll want a RWMutex here. Or we could probably use our cache in gotools as well if you'd want to open source it

Reply:
Good catch! I need to get into a better habit of running with `-race` (though in this case the simple existing test didn't trigger, added a test that would)